catalyst_voices_packages/catalyst_cardano_serialization/lib/src/witness.dart

```diff
@@ -1,35 +1,161 @@
+  vkeyWitness(0),
+
+  /// A native script witness.
+  nativeScript(1),
+
+  /// A bootstrap witness.
+  bootstrapWitness(2),
+
+  /// A Plutus v1 script witness.
+  plutusV1Script(3),
+
+  /// A Plutus data witness.
+  plutusData(4),
+
+  /// A redeemer witness.
+  redeemers(5),
+
+  /// A Plutus v2 script witness.
+  plutusV2Script(6),
+
+  /// A Plutus v3 script witness.
+  plutusV3Script(7);
+
+  /// The integer value associated with the witness type.
+  final int value;
+
+  /// Constructs a new `WitnessType` with the given integer value.
+  const WitnessType(this.value);
+}
+
 /// A set of witnesses that sign the transaction.
 final class TransactionWitnessSet extends Equatable implements CborEncodable {
   /// The witnesses that sign the transaction.
   final Set<VkeyWitness> vkeyWitnesses;
 
+  /// The native scripts.
+  final Set<NativeScript> nativeScripts;
+
+  /// The Plutus V1 scripts.
+  final Set<PlutusV1Script> plutusV1Scripts;
+
+  /// Redeemers for Plutus scripts.
+  // FIXME(ilap): implement proper redeemers type.
+  final Set<Redeemer> redeemers;
+
+  /// The Plutus V2 scripts.
+  final Set<PlutusV2Script> plutusV2Scripts;
+
+  /// The Plutus V3 scripts.
+  final Set<PlutusV3Script> plutusV3Scripts;
+
   /// The default constructor for [TransactionWitnessSet].
-  const TransactionWitnessSet({required this.vkeyWitnesses});
+  const TransactionWitnessSet({
+    this.vkeyWitnesses = const {},
+    this.nativeScripts = const {},
+    this.plutusV1Scripts = const {},
+    this.redeemers = const {},
+    this.plutusV2Scripts = const {},
+    this.plutusV3Scripts = const {},
+  });
 
   /// Deserializes the type from cbor.
   factory TransactionWitnessSet.fromCbor(CborValue value) {
     final map = value as CborMap;
     return TransactionWitnessSet(
-      vkeyWitnesses: map.values.map(VkeyWitness.fromCbor).toSet(),
+      vkeyWitnesses: _getWitnesses<VkeyWitness>(
+        map,
+        WitnessType.vkeyWitness,
+        VkeyWitness.fromCbor,
+      ),
+      nativeScripts: _getWitnesses<NativeScript>(
+        map,
+        WitnessType.nativeScript,
+        NativeScript.fromCbor,
+      ),
+      plutusV1Scripts: _getWitnesses<PlutusV1Script>(
+        map,
+        WitnessType.plutusV1Script,
+        PlutusV1Script.fromCbor,
+      ),
+      redeemers: _getWitnesses<Redeemer>(
+        map,
+        WitnessType.redeemers,
+        Redeemer.fromCbor,
+      ),
+      plutusV2Scripts: _getWitnesses<PlutusV2Script>(
+        map,
+        WitnessType.plutusV2Script,
+        PlutusV2Script.fromCbor,
+      ),
+      plutusV3Scripts: _getWitnesses<PlutusV3Script>(
+        map,
+        WitnessType.plutusV3Script,
+        PlutusV3Script.fromCbor,
+      ),
     );
   }
 
+  static Set<T> _getWitnesses<T>(
+    CborMap map,
+    WitnessType type,
+    T Function(CborValue) fromCbor,
+  ) {
+    final value = map[CborSmallInt(type.value)];
+    return value != null ? (value as CborList).map(fromCbor).toSet() : {};
```

Comment:
Seems as more dart-like approach: 

```suggestion
    final value = map[CborSmallInt(type.value)] as CborList?;
    return value?.map(fromCbor).toSet() ?? {};
```